### DISCUSSION TITLE
Calculation of SOC in the driver by Coloumb-Counting

### DISCUSSION THREAD

**Discussion by cflenker** [Louisvdw/dbus-serialbattery#869] [Show and tell]
Thank you very much @Louisvdw and @mr-manuel for your great work of implementing this great project!
I am using this driver at my home and in 5 other victron 3-phase installations in combination with jkbms.

A big disadvantage of the jkbms ist, that you can not configure a soc-calibration-voltage. The only possibility to set the soc to 100% is to hit the OVP-voltage. That means the bms switches off the charging-fets. 

My suggestion would be to do the coloumb-counting in the driver and build an own SOC in the driver. Then you have the possibility to configure the reset-condition on your own and maybe you do not need to charge to a high bulk-voltage every few days.

I Implemented this on my own fork of the projekt and it works fine for me (https://github.com/Louisvdw/dbus-serialbattery/pull/868).

Here you can see the calculated SOC of the driver compared to the SOC of the victron smartshunt:
![image](https://github.com/Louisvdw/dbus-serialbattery/assets/125555670/8e8539d0-757b-42c2-bc26-b271099a8ee2)

Here are the conditions to reset the SOC to 100%:

> ; --------- SOC calculation ---------
> ; Description: Calculate the SOC in the driver. Do not use the SOC reported by the BMS
> ; SOC_CALCULATION = True: Calc SOC in the driver, do not use SOC reported from BMS
> ;      The SOC is calculated by integration of the current reported by the BMS.
> ;      The SOC is set to 100% if following conditions apply for at least SOC_RESET_TIME seconds:
> ;          * current is lower than SOC_RESET_CURRENT Amps
> ;          * Sum of Cell voltages >= MAX_CELL_VOLTAGE * Cell Count - VOLTAGE_DROP
> ; SOC_CALCULATION = False: Use SOC reported from BMS
> SOC_CALCULATION = False
> SOC_RESET_CURRENT = 7
> SOC_RESET_TIME = 60

**Comment by mr-manuel** [comment]
> A big disadvantage of the jkbms ist, that you can not configure a soc-calibration-voltage. The only possibility to set the soc to 100% is to hit the OVP-voltage. That means the bms switches off the charging-fets.

Yes, this is possible in the current `dev` branch.

I like the idea. I will check what side effects this could have and test the driver. This will take some time.
